### ISSUE TITLE
Bumblebee for Desktop

### ISSUE THREAD

**Issue by Bugattikid2012** [Bumblebee-Project/Bumblebee#577]
$ optirun
[  641.107275] [ERROR]The Bumblebee daemon has not been started yet or the socket path /var/run/bumblebee.socket was incorrect.
[  641.107374] [ERROR]Could not connect to bumblebee daemon - is it running?

That is the error I recieve.  I've tried re-installing nvidia and bumblebee countless times, I've tried different ways of installing and everything (though I haven't tried a manual install, but it probably won't make a difference).  To sum it up, I've done everything on this forum and more.

http://askubuntu.com/questions/454729/installed-14-04-and-nvidia-graphics-card-wont-load

I've look and looked on the internet, and it seems that nothing will work for me.  The only solution I can think of is if I can get a version before 3.1 of this project.  I've tried using the terminal install listed here

http://www.howtogeek.com/117929/how-to-downgrade-packages-on-ubuntu/

but it seems only the latest version is available.  I NEED this to work, as I can't even get my main monitor to work without it.  Please help!  Thanks.

Specs:

MSI 780 Ti Gaming

i7 4770k

16 gigs ram

SSD for boot and games that need it HDD for storage

Dual booting Ubuntu 14.04 with Windows 7 (or Winderp$ as I like to call it)

**Comment by ArchangeGabriel** [comment]
Bumblebee can’t work on desktop, because bbswitch won’t load. We might consider making bbswitch an optdepend, but I’m not sure it’s worth it.

**Comment by Bugattikid2012** [comment]
How much effort would it actually be?  I don't understand what exactly is holding you back, there's no such thing as a different OS for a laptop vs for a desktop.  They function the same way, unless Linux is just weird like that.  Laptop's aren't lacking anything desktops have, they're just smaller and are designed to use less power.

**Comment by ArchangeGabriel** [comment]
The difference is on BIOS/ACPI side. Laptop have specific functions there for Optimus that desktop don’t have, and we use them in bbswitch.

What happen in your case must be that Bumblebee starts, loads bbswitch, bbswitch try to find those functions, fail to do so, so fail to load, which leads Bumblebee to abort loading — thus explaining why you can’t connect to the daemon.

I’m not sure that’s a lot of effort, we must mainly change how Bumblebee depends on bbswitch to make it optional, but care about reporting appropriate error if bbswitch failed to load on a laptop, and about packaging so that bbswitch is still a default for laptop while not being installed for desktop (I suppose that for this part, making a bumblebee-desktop package is the right way).

What is holding me is more that, like @amonakov, I’m not sure Bumblebee is the right tool for this, and perhaps it already exists something that handle this case of hybrid GPU for desktop. I need to find documentation about this.

**Comment by Bugattikid2012** [comment]
Thank you for your interest.  I also wasn't expecting BIOS/UEFI to be the limiting factor here, but that does make more sense.  

I guess I should clarify what I am attempting to do.  My plan was/is to have my main monitor running from my dedicated card, and my second monitor running from my integrated card.  There is a noticeable difference when my dedicated card is not running two monitors, and just one.  It stops a bit of stuttering and sometimes results in upwards of 15 FPS differences.  Also, if my dedicated card is getting noticeably hot, I can run off of my integrated card and second monitor for a few minutes while I wait for my dedicated card to cool off.  It won't be doing much of anything if the computing is going on at the integrated card, which results in a noticeably faster cool down time.  

It sounds like this is going to the extreme details, and admittedly it sort of is.  However, I'm not talking about 1 FPS difference, it's a nice difference that results in a decent 15+ FPS boosts, so I think it's worth it.  The cool down time isn't that much, but it's still something.  

I understand why you were confused with my request, I always knew bumblebee was used for laptops mainly, but I didn't realize they didn't work at all on desktops (usually).  I haven't been able to find any alternatives to bumblebee, but if I do find one I will try to post it here.  Thank you for your interest and help.  If you have anything else that could be of assistance to my issue, it would be greatly appreciated if you alerted me about it.

**Comment by ArchangeGabriel** [comment]
OK, this is in fact far from what we achieve here. You’re going to have your NVIDIA card powered all the time, one screen on it, one screen on the Intel.

After a little looking around this, even on Windows, Intel relies on a paid software called LucidVertu that they used to bundle with some motherboards, and this software let you decide which card is managing screen(s) (even if it’s not the card where they are plugged in, but in this case at a high cost of performances) and which card is rendering each application. But you can’t run one screen on a card, one screen on the other one. And test around there of LucidVertu says it’s a crap.

What is your setup on Windows?

On Linux, does it work if you plug one screen on each? I mean, are both screens detected and used in this case? I would be very surprised of that.

Here, in Bumblebee multi-monitor setups (that concerns users for which the HDMI output is wired to the NVIDIA chip), you have to start a second X server on the external monitor, and there you can start an other DE/Window Manager, but it’s not very stable, and most important point, since this isn’t the same server/DE/WM, you can’t move window from one screen to the other.

Through PRIME (and its reverse setup), you can choose (ideally, IRL all the drivers, especially NVIDIA, are not working in all modes) which card manage the screens, and which one render each application, but here again you can’t have one screen on each card.

So, AFAIK, there is no existing solution with the two screens being managed each by a different card while still being part of the same desktop, I’m not sure this is technically possible, but even if it is, it must be an huge effort to implement, with quite no interest.

I think you should either let your integrated card manage both screen while doing some rendering using the NVIDIA one (this could be achieved using PRIME maybe, or at least a desktop version of Bumblebee), start an X server on each screen with a different card, each with a DE/WM, but this must be a lot of pain to setup this, or just let your NVIDIA card do everything at this cost of 15 FPS (as long as you’re having 60 or even 30, you’re fine I think).

**Comment by Bugattikid2012** [comment]
Thanks for the detailed reply.  I'm currently running a 144hz monitor, so the higher the settings while still remaining 144ish FPS is key.  (Please don't start talking about some BS about the human eye can't see higher than 30FPS, it's SOOOOO incorrect.  If you don't believe me, please kindly ask for some sources and I'll be more than happy to fetch some for you.  Also, the eye can benefit from as many FPS as you can throw at it, however there won't be any /significant/ benefit after 75~-150~ FPS, depending on each and every individual.  I can vouch that 144 FPS/hz makes the world of difference, even when just moving your mouse around on a desktop.)

On Winderp$, I can run an AMD card on one monitor, and an NVIDIA card on the other.  Heck, I could even run one from AMD, one from NVIDIA, and a third from Intel (in theory at least, I don't know if it's actually been done, but I don't see anything limiting it on Winderp$.  I can confirm that any combination of two of the three cards will work fine.).  

My setup on Winderp$ is exactly as I want it to be on Linux.  I'm running my main monitor out of Dual Link DVI-D from my NVIDIA card, and my second one is hooked up through HDMI to the Intel port from my motherboard.  It's never given me issues, at least not on this PC.  I've had a few issues on my laptop that runs the same way (intel 4000 series HD graphics chip, and a 650M from NVIDIA), but it was resolved pretty easily.  Winderp$ update had automatically installed a driver for my Intel chip, and something went wrong.  After a little troubleshooting, I found out all I had to do was reinstall the Intel graphics driver.  So no /real/ issues on Winderp$.  

On Linux, if I have my integrated graphics enabled in my BIOS/UEFI, it's the only thing Linux will detect, and it refuses to admit my NVIDIA card exists.  If I disable it, I get some potato resolution around 640x480 and I can't change it very easily.  After some tinkering, I can get it to work though.  Then, once I enable my integrated card again, Linux prefers it over my dedicated card, despite my BIOS/UEFI settings to use my dedicated card if possible.  

I guess I'll just have to leave it as it is for the moment, I haven't even used Linux in months because of this issue.  

I don't remember their names, but supposedly there's two new things that are out in the works to replace X-org.  I think one of them starts with an M, and Canonical is sort of spear heading that one along if I recall correctly.  These obviously won't be perfect for a few years to come, but is it /PROBABLE/ that these things could fix my issue?  Will they give more performance?  Less bugs?  Etc?  And how soon until these are estimated to be perfected/patched?  

I guess I'll just wait until these two come out and see what they have to offer.  Thanks again for your very dedicated help.

**Comment by ArchangeGabriel** [comment]
OK, didn’t know M$ was having this working, I would be interested to know how they do that.

Currently, AFAIK, this is not possible under Linux, but maybe one day…

The two things you mention are Mir from Canonical, which was expected to land in 13.10 at first, and isn’t expected to land before 16.04 now, so I suppose this will be 18.04 or something like that, and the second one in Wayland. Back when Wayland devel started, people where expecting it to be in use around 2015, but I don’t know what is the state of the project today and know no-one using it, plus I have no idea on how they handle things like multi-GPU and/or multi-monitor.

And in both cases, I think it’s to early to talk about performance and bugs, because I expect both to be not as reliable as X.org today, but the goal of reimplementing a display server is more performance, less bugs, and better maintaining/evolving.

And in 2010, PRIME was expected to work fully in 2012, and today it’s just almost working. So, be prepare to wait a little more I think before being able to use such configuration under Linux.

**Comment by Bugattikid2012** [comment]
Wow, it got pushed that far back?  I was expecting it to remain /fairly/ close to the original release dates.  But 18.04?  Dang.  Good thing I don't use Ubuntu anymore.  

I guess we'll just have to wait and see if Wayland comes out with a fix for this, and that it comes out in the next decade...  Stupid Xorg... Relevant XKCD: http://xkcd.com/963/  Don't forget the alt text!

Also this one: http://xkcd.com/456/

Thanks for your help.  I'm sure one day I'll find/make a fix if one doesn't exist already by then, but I don't have the knowledge to do either at the moment.  If I ever find one I'll let you know.

**Comment by ArchangeGabriel** [comment]
Well, they say it will be 16.04. But the past teached us that every time they give a deadline like this, it’s next cycle.

**Comment by wilfm** [comment]
Nearly the same for Wayland as well....

Anyway, are there any desktop boards that implement BIOS/ACPI similar to laptops, so that Bumblebee (or any other hybrid graphics optimisation stuff) will work? (sorry I'm a bit of a noob in this area :)

**Comment by Bugattikid2012** [comment]
I'm not sure about boards, but I recently found out how [Arch Linux claims to work with Bumblebee.  ](https://wiki.archlinux.org/index.php/Bumblebee)

**Comment by wilfm** [comment]
Ok :). Just I'm interested due to a need to upgrade my desktop (but don't want it to use loads of power all the time), and I have found that some desktop boards seem to have a some weird graphics implementations available (e.g. LucidLogiz Virtu: http://www.gigabyte.com/MicroSite/279/images/mb-z68-built-in-visuals.html), but possibly under Windowz only.

With NVIDIA's 'Optimus', they apparently were going to launch for desktop (http://news.softpedia.com/news/NVIDIA-Optimus-Lands-on-Desktops-196761.shtml) but never did.

**Comment by dvarum12** [comment]
I can confirm that it worked under ArchLinux, but it will stopped working after I put the computer to sleep and resumed, then I had to reboot the computer and it was working again. However it stopped working for me at some point, not sure then and that is way I am reading this Issue. 
I don't know if it works for other Arch users, but it was nice then it worked for me, because I was saving about 80-100W.

**Comment by ArchangeGabriel** [comment]
@dvarum12 Is there any chance you still have logs from that time where it worked? If you have a high disk usage and time limit for journalctl, it might (I’m still having full logs from my initial install), in which case we could see what changed, and if there is something we could do about it. Also, how did you measure that power saving?

**Comment by dvarum12** [comment]
@ArchangeGabriel I dont have the old logs, but I do have it working again, with steam on ArchLinux :smile:

**Comment by ArchangeGabriel** [comment]
So, you’re currently having your screen handled by the integrated Intel card, use Bumblebee to run things on the Nvidia card, and get power savings in this situation vs. using the Nvidia card for the display? No bbswitch involved, right?

**Comment by dvarum12** [comment]
@ArchangeGabriel yes, and considering that I am not a big gamer, but do playing from time to time. This is a nice thing to have. Of course, this is not perfect and there do exist problems, but for me the pros out ways the cons.

**Comment by ArchangeGabriel** [comment]
Could you give a dmesg log, and Xorg.{0,8}.log too? That sounds interesting.

**Comment by dvarum12** [comment]
Yes, then I get home I will get that for you.

**Comment by dvarum12** [comment]
As promised, hope you find something :smiley: 
[dmesg.txt](https://github.com/Bumblebee-Project/Bumblebee/files/257014/dmesg.txt)
[Xorg.0.log.txt](https://github.com/Bumblebee-Project/Bumblebee/files/257020/Xorg.0.log.txt)
[Xorg.8.log.txt](https://github.com/Bumblebee-Project/Bumblebee/files/257019/Xorg.8.log.txt)

**Comment by ArchangeGabriel** [comment]
OK, thanks. So, Bumblebee does indeed work for desktop as a free bonus since we added support for different PM backend, none being one of the option (thus no failure with bbswitch unable to load, which was the point in this early discussion), however I see no evidence of real powersaving here. We would indeed need a desktop equivalent to Optimus at ACPI level I think for that, unless we can manage perhaps to turn of the PCIe port (not sure whether this is doable for desktop, especially without putting the card off first, but there is at least the idea: https://github.com/Bumblebee-Project/bbswitch/pull/130).

Or maybe there’s some sort of magic somewhere that already do this automatically, hence the power savings your report (that or you didn’t compare the right things).

I’m not sure if this is easy for you to try (you will have to change a bit some config files at least), but it would be interesting to see power consumption when idling with screen wired to Intel vs Nvidia, and also perfs in both cases.

**Comment by dvarum12** [comment]
So you don't be leave in magic :(
My guess is that the graphics card goes into some kind of sleep mod then it is not used, but I don't know.
I don't have to much time to play around with it right now, but then school is over I will be happy to do some testing in it for you. If you are interested?

**Comment by ArchangeGabriel** [comment]
Neh, I’m a scientist. :p All things have an explanation, including the explanations themselves, except maybe at the end of the chain, but that’s too philosophical. ;)

Yes, sure that would be interesting. Looking forward to hearing again from you soon. :)

**Comment by dvarum12** [comment]
Hey @ArchangeGabriel school is over and I am ready to mess with some config files :smile:

**Comment by sabian2008** [comment]
Just wanted to report I might be able to do some tests too. I've just installed bumblebee in Debian Testing and it worked out of the box (with lightdm, gdm3 won't start but don't have time to look into that now). My goal was to test power consumption (and to play with something new on a boring day). How can I monitor power consumption?

I have a NVidia 750Ti and a " VGA compatible controller: Intel Corporation Xeon E3-1200 v3/4th Gen Core Processor Integrated Graphics Controller (rev 06)" Intel Card.

I don't know the potential drawbacks of powering off the card could be (using acpi_call?). Any insights?

**Comment by Lekensteyn** [comment]
@sabian2008 The power saving methods are not available for desktop, only "Optimus" laptops with Hybrid graphics technology. bbswitch, `acpi_call`, etc. won't work because the ACPI methods are missing.

If you want to offload graphics rendering, you could have a look at "DRI PRIME", but do not expect huge benefits here. Just plug your monitor in the stronger GPU.

**Comment by sabian2008** [comment]
@Lekensteyn Yes, after I posted I noticed the blog post I had superficially read it used acpi_call on desktop, had an error in the introduction and the rest of the article talked about laptops.

Are there any potential drawbacks of using bumblebee without bbswitch on a desktop? I only play demanding games about 6 hours per month and sometimes I use [Paraview.](http://www.paraview.org/) In neither case I've noticed appreciable slowdowns running with `primusrun`. So I guess it seems to work for me and keep my NVidia card at very low power levels and fan speed (1W and 0% respectively, according to `nvidia-sli`, I don't have hardware to measure that stuff). 

My only concern (as I'm ignorant of the low level stuff) is if it could damage the hardware (intuitively I'd say it's impossible).

Sorry if I posted off-topic here, didn't mean to.

Cheers,

**Comment by Lekensteyn** [comment]
@sabian2008 Bumblebee can indeed be used without bbswitch, it won't damage your hardware. No problem to ask here, it seems on-topic

**Comment by rokups** [comment]
I am interested in this as well. Is it in any way possible to run this with nvidia libgl installed? I am aware that bumblebee requires integrated GPU to use mesa libraries, however my main gpu is old nvidia card which freezes with noveau driver. Since i am not aware of bumblebee design could any of developers comment if it is in theory possible to get bumblebee working with nvidia libgl as main libgl? Any pointers what part of code i could poke to make it happen?

**Comment by Lekensteyn** [comment]
@rokups You have both an integrated and discrete Nvidia GPU, or what is the situation? What is your goal?

**Comment by rokups** [comment]
I have two discrete GPUs. One is used for passthrough for VM sometimes, one is used for host. Goal is to use bumblebee to start some 3d applications on VM GPU (when it is not in use by VM ofc) as it is more powerful.

**Comment by ArchangeGabriel** [comment]
@rokups Not sure how much stronger is your second GPU, but Bumblebee has a very high overread (and even more without Primus, VirtualGL being the only likely to work setup for you), so this might a thing…

**Comment by rokups** [comment]
Well it is gtx630 vs gtx1080 so.. I would still like to try. Why wouldnt primus work?

Also i got package installed and tried to start `optirun glxgears`. Strange thing happens. Log contains:
```
[   544.086] (++) Using config file: "/etc/bumblebee/xorg.conf.nvidia"
[   544.086] (**) |   |-->Device "DiscreteNvidia"
```
And i added `BusID  "PCI:5:0:0"` to `xorg.conf.nvidia`. It should be using that card. But later in the log:
```
[   544.087] (II) xfree86: Adding drm device (/dev/dri/card0)
[   544.089] (--) PCI:*(0:3:0:0) 10de:0f00:1458:3544 rev 161, Mem @ 0xde000000/16777216, 0xc8000000/134217728, 0xd0000000/33554432, I/O @ 0x0000e000/128, BIOS @ 0x????????/131072
```
Now that is my primary GPU. Naturally it fails:
```
[   544.093] (EE) NVIDIA(GPU-0): Failed to initialize the NVIDIA graphics device!
```
Any idea whats going on here? Attached full log as well.
[Xorg.8.log.txt](https://github.com/Bumblebee-Project/Bumblebee/files/758612/Xorg.8.log.txt)

**Comment by Lekensteyn** [comment]
@rokups I don't think that Bumblebee is appropriate for your scenario. You presumably still have video outputs on your more powerful GPU. One option is then to plug in your monitor in that GPU. It will be a more pleasant experience than Bumblebee I think. Bumblebee currently relies on unloading modules to unbind the driver from the GPU device. If both your primary and secondary GPUs use the same driver, then you have a problem.

[PRIME GPU offloading][1] is another option, but this is not supported by Nvidia blob.

 [1]: https://wiki.archlinux.org/index.php/PRIME#PRIME_GPU_offloading

**Comment by rokups** [comment]
Still does not sound like impossible task. Currently i am using a shellscript that binds driver to the card, starts X, after X quits it binds `vfio-pci` driver to the card. Nothing that cant be done with some bumblebee patching i guess.

**Comment by deveee** [comment]
@ArchangeGabriel  It's not just for power saving. It would also make a computer more silent and it would make a possibility to easily test applications on different hardware (easy switch between intel and nvidia). And all of them are IMO useful.
